keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1066,6 +1076,264 @@ the remaining requests in that queue start getting faster service.  In
 both cases, the service delivery in the virtual world has reacted
 properly to the true service duration.
 
+### Support for LIST requests
+
+Up until now, we were assuming that even though the requests aren't
+necessarily equally expensive, their actual cost is actually greatly
+reflected by the time it took to process them.  But while being processed
+each of them is consuming the equal amount of resources.
+
+It works well for requests that are touching only a single object.
+However, given the fact that in practise the concurrency limits has to be
+set much higher than number of available cores to achieve reasonable system
+throughput, this no longer works that well for LIST requests that are orders
+of magnitude more expensive.  The hand-wavy way of rationalizing it is that
+the ratio of time the request is processed by the processor to the actual
+time of processing the request starts to visibly differ (e.g. due to I/O
+waiting time - there is communication with etcd in between for example).
+
+As a result, kube-apiserver (and etcd) may be able to easily keep with N
+simple in-flight requests (e.g. create or get a single Pod), but will explode
+trying to process N requests listing all the pods in the system at the same
+time.
+
+#### Width of the request
+
+In order to address this problem, we are introducing the concept of `width`
+of the request.  Instead of saying that every request is consuming a single
+unit of concurrency, we allow for a request to consume `<width>` units of
+concurrency while being processed.
+
+While in theory the `width` can be an arbitrary non-integer number, for
+practical reasons, we will assume it actually is an integer.  Given that
+our estimations here are very rough anyway that seems a reasonable
+simplification that makes dispatching the budget a bit simpler.
+
+#### Determining the width
+
+While one can imagine arbitrarily sophisticated algorithms for it (including
+exposing defining the width of requests via FlowSchema API), we are going
+to start with something simple to first get operational experience with it.
+
+While the actual cost of processing a LIST request is roughly proportional
```

Comment:
I think this paragraph is talking about CPU, but if you made it also/instead about RAM, the analogy would be more exact. Since we don't stream results and RAM actually is more or less proportional to the number of items.

Reply:
Indeed the RAM makes it proportional to the number of items.
That said, that could mean that we want the "width" to be set proportionally to number of items, which I think it shouldn't.

I think the argument I'm trying to make in general (which isn't yet well describe here - will update) is that the cost of the request is now `<width> x <latency>`. The question is how to set width. I think the parallelization argument is a good one, so I'm not entirely sure we should really change here (other than that I need to explain that the cost is actually `width x latency`.

Reply:
> That said, that could mean that we want the "width" to be set proportionally to number of items, which I think it shouldn't.

I wrote a really long response agreeing and trying to come up with math to justify this intuition but changed my mind by the end.

Assume the true cost of a request is multi-dimensional e.g. `{RAM, CPU} * duration`, and ram is proportional to the number of items but CPU is relatively constant (the request takes longer with more items but the CPU isn't working harder over that time). Then the CPU actually doesn't matter much for scheduling, since all requests consume the same amount, and the only thing that determines the width is the RAM cost.

This clearly implies that it should be proportional, yet that still seems quite wrong?

One way to rescue this might be to note that RAM usage actually starts at zero and goes up as results from etcd come in, and then stays there until we're finished writing to the client. If we average that out for N items, it'd be something like N(N+1)/2*duration. But duration is probably also a function of N, so that reduces yet again to the estimate being linear in N, it's just scaled down.

So I guess I think we should probably try width = N * (scale factor). Every approach I tried (not all written down here) gave me a linear estimator.

Reply:
I spend some time thinking about it and I think we were both right and wrong at the same time.
I was too focused on CPU, you were too focused on RAM. And we were both right about those dimensions.
But we didn't think about combining them together correctly.

I rewrote this section to reflect that - hopefully it makes more sense no.